ios/KeyCommand.mm

```diff
@@ -1,31 +1,341 @@
+}
+
+- (void)handleKeyUIEventSwizzle:(UIEvent *)event
+{
+  NSString *unmodifiedInput = nil;
+  NSString *modifiedInput = nil;
+  UIKeyModifierFlags modifierFlags = 0;
+  BOOL isKeyDown = NO;
+  long keyCode = 0;
+
+  if ([event respondsToSelector:@selector(_unmodifiedInput)]) {
+    unmodifiedInput = [event _unmodifiedInput];
+  }
+
+  if ([event respondsToSelector:@selector(_modifiedInput)]) {
+    modifiedInput = [event _modifiedInput];
+  }
+
+  if ([event respondsToSelector:@selector(_modifierFlags)]) {
+    modifierFlags = [event _modifierFlags];
+  }
+
+  if ([event respondsToSelector:@selector(_isKeyDown)]) {
+    isKeyDown = [event _isKeyDown];
+  }
+
+  if ([event respondsToSelector:@selector(_keyCode)]) {
+    keyCode = [event _keyCode];
+  }
+
+  BOOL interactionEnabled = !UIApplication.sharedApplication.isIgnoringInteractionEvents;
+  BOOL hasFirstResponder = NO;
+  if (isKeyDown && modifiedInput.length > 0 && interactionEnabled) {
+    UIResponder *firstResponder = nil;
+    for (UIWindow *window in [self allWindows]) {
+      firstResponder = [window valueForKey:@"firstResponder"];
+      if (firstResponder) {
+        hasFirstResponder = YES;
+        break;
+      }
+    }
+
+    // Ignore key commands (except escape) when there's an active responder
+    if (!firstResponder) {
+      [self RCT_handleKeyCommand:unmodifiedInput flags:modifierFlags];
+    }
+  }
+};
+
+- (NSArray<UIWindow *> *)allWindows
+{
+  BOOL includeInternalWindows = YES;
+  BOOL onlyVisibleWindows = NO;
+
+  // Obfuscating selector allWindowsIncludingInternalWindows:onlyVisibleWindows:
+  NSArray<NSString *> *allWindowsComponents =
+      @[ @"al", @"lWindo", @"wsIncl", @"udingInt", @"ernalWin", @"dows:o", @"nlyVisi", @"bleWin", @"dows:" ];
+  SEL allWindowsSelector = NSSelectorFromString([allWindowsComponents componentsJoinedByString:@""]);
+
+  NSMethodSignature *methodSignature = [[UIWindow class] methodSignatureForSelector:allWindowsSelector];
+  NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
+
+  invocation.target = [UIWindow class];
+  invocation.selector = allWindowsSelector;
+  [invocation setArgument:&includeInternalWindows atIndex:2];
+  [invocation setArgument:&onlyVisibleWindows atIndex:3];
+  [invocation invoke];
+
+  __unsafe_unretained NSArray<UIWindow *> *windows = nil;
+  [invocation getReturnValue:&windows];
+  return windows;
+}
+
++ (instancetype)voidCompletionBlock
+{
+  return nil;
+}
+
+- (void)RCT_handleKeyCommand:(NSString *)input flags:(UIKeyModifierFlags)modifierFlags
+{
+  for (HardwareKeyCommand *command in [HardwareKeyCommands sharedInstance].commands) {
+    if ([command matchesInput:input flags:modifierFlags]) {
+      if (command.block) {
+        UIKeyCommand *keycommand = [UIKeyCommand keyCommandWithInput:input modifierFlags:modifierFlags action:@selector(voidCompletionBlock:)];
+        command.block(keycommand);
+      }
+    }
+  }
+}
+
++ (instancetype)sharedInstance
```

Comment:
singleton implementation should be one of the first being declared on the class, along with the init

Reply:
done